// appium-xcuitest-driver/build/lib/ios-deploy.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var _appiumSupport = require('appium-support');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var IOSDEPLOY_PATH = 'ios-deploy';

var IOSDeploy = (function () {
  function IOSDeploy(udid) {
    _classCallCheck(this, IOSDeploy);

    this.udid = udid;
    this.cmd = IOSDEPLOY_PATH; // this.cmd is in accordance with iDevice
  }

  _createClass(IOSDeploy, [{
    key: 'checkStatus',
    value: function checkStatus() {
      return _regeneratorRuntime.async(function checkStatus$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.which(this.cmd));

          case 2:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'remove',
    value: function remove(bundleid) {
      var remove;
      return _regeneratorRuntime.async(function remove$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            remove = ['--uninstall_only', '--id', this.udid, '--bundle_id', bundleid];
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)(this.cmd, remove, { maxBuffer: 524288 }));

          case 4:
            context$2$0.next = 10;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](1);

            _logger2['default'].debug('Error : ' + context$2$0.t0.message);
            throw new Error('Could not remove app ' + context$2$0.t0.message);

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 6]]);
    }
  }, {
    key: 'install',
    value: function install(app) {
      var install;
      return _regeneratorRuntime.async(function install$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            install = ['--id', this.udid, '--uninstall', '--bundle', app];
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)(this.cmd, install, { maxBuffer: 524288 }));

          case 4:
            context$2$0.next = 10;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](1);

            _logger2['default'].debug('Error : ' + context$2$0.t0.message);
            throw new Error('Could not install app ' + context$2$0.t0.message);

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 6]]);
    }
  }, {
    key: 'isInstalled',
    value: function isInstalled(bundleid) {
      var isInstalled, _ref, stdout;

      return _regeneratorRuntime.async(function isInstalled$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            isInstalled = ['--exists', '--id', this.udid, '--bundle_id', bundleid];
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)(this.cmd, isInstalled, { maxBuffer: 524288 }));

          case 4:
            _ref = context$2$0.sent;
            stdout = _ref.stdout;
            return context$2$0.abrupt('return', stdout && stdout.indexOf("true") > -1);

          case 9:
            context$2$0.prev = 9;
            context$2$0.t0 = context$2$0['catch'](1);

            _logger2['default'].debug('Error checking install status: ' + context$2$0.t0.message);
            return context$2$0.abrupt('return', false);

          case 13:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 9]]);
    }
  }]);

  return IOSDeploy;
})();

exports['default'] = IOSDeploy;
module.exports = exports['default'];

// make sure we actually have the program
